server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -481,12 +546,17 @@ public void rejectedExecution(Runnable task, ThreadPoolExecutor executor) {
             }
         }
 
-        private static void put(ThreadPoolExecutor executor, Runnable task) {
+        private void put(ThreadPoolExecutor executor, Runnable task) {
             final BlockingQueue<Runnable> queue = executor.getQueue();
-            // force queue policy should only be used with a scaling queue
-            assert queue instanceof ExecutorScalingQueue;
+            // force queue policy should only be used with a scaling queue (ExecutorScalingQueue / LinkedTransferQueue)
+            assert queue instanceof LinkedTransferQueue;
             try {
                 queue.put(task);
+                if (probeWorkerPool && task == queue.peek()) { // referential equality
+                    // If the task is at the head of the queue, we can assume the queue was previously empty. In this case available workers
+                    // might have timed out in the meanwhile. To prevent the task from starving, we submit a noop probe to the executor.
```

Comment:
Mmmm... (thinking out loud) this might be more robust than  `(executor.getPoolSize() == 0)` ; it does not assume there might be a worker, just that the queue was empty.
But testing for executor size should still be OK? I mean, if you have 1 worker at this point it's going to pick up the queue? Maybe not, there could be a race here if using this condition.
I think you made the right call with the queue size instead, and I wonder if this is a bug in the JDK (they do test for worker count IIRC!)

Reply:
yes, i'm deliberately using the queue head here... probably being paranoid.
Tests pass if using `executor.getPoolSize() == 0`, but I'm afraid `workers.size()` doesn't reflect the atomic size of workers returned by `workerCountOf(c)`. `processWorkerExit` will first decrement the atomic size, before actually removing a worker from the `worker`.

Reply:
> probably being paranoid

I don't think so, it seems reasonable and more robust, so ++ for me

Reply:
will add as a comment